File: src/bmsx/machine/devices/irq/controller.ts
import { IO_IRQ_ACK, IO_IRQ_FLAGS } from '../../bus/io';
import type { Value } from '../../cpu/cpu';
import { Memory } from '../../memory/memory';

export type IrqControllerState = {
	pendingFlags: number;
};

export class IrqController {
	private pendingFlags = 0;

	public constructor(private readonly memory: Memory) {
		this.memory.mapIoRead(IO_IRQ_FLAGS, this.onFlagsRead.bind(this));
		this.memory.mapIoWrite(IO_IRQ_ACK, this.onAckWrite.bind(this));
	}

	public reset(): void {
		this.pendingFlags = 0;
		this.clearAckLatch();
	}

	public postLoad(): void {
		this.clearAckLatch();
	}

	public captureState(): IrqControllerState {
		return {
			pendingFlags: this.pendingFlags,
		};
	}

	public restoreState(state: IrqControllerState): void {
		this.pendingFlags = state.pendingFlags >>> 0;
		this.clearAckLatch();
	}

	public hasAssertedMaskableInterruptLine(): boolean {
		return this.pendingFlags !== 0;
	}

	public raise(mask: number): void {
		const next = (this.pendingFlags | (mask >>> 0)) >>> 0;
		if (next !== this.pendingFlags) {
			this.pendingFlags = next;
		}
	}

	private onFlagsRead(): Value {
		return this.pendingFlags;
	}

	private onAckWrite(_addr: number, value: Value): void {
		const ack = (value as number) >>> 0;
		if (ack !== 0) {
			const next = (this.pendingFlags & ~ack) >>> 0;
			if (next !== this.pendingFlags) {
				this.pendingFlags = next;
			}
		}
		this.clearAckLatch();
	}

	private clearAckLatch(): void {
		this.memory.writeIoValue(IO_IRQ_ACK, 0);
	}
}

File: src/bmsx_cpp/machine/devices/irq/controller.cpp
#include "machine/devices/irq/controller.h"
#include "machine/bus/io.h"

namespace bmsx {

IrqController::IrqController(Memory& memory)
	: m_memory(memory) {
	m_memory.mapIoRead(IO_IRQ_FLAGS, this, &IrqController::onFlagsReadThunk);
	m_memory.mapIoWrite(IO_IRQ_ACK, this, &IrqController::onAckWriteThunk);
}

void IrqController::reset() {
	m_pendingFlags = 0;
	m_memory.writeIoValue(IO_IRQ_ACK, valueNumber(0.0));
}

void IrqController::postLoad() {
	const Value clearAck = valueNumber(0.0);
	m_memory.writeIoValue(IO_IRQ_ACK, clearAck);
}

void IrqController::raise(uint32_t mask) {
	const uint32_t next = m_pendingFlags | mask;
	if (next != m_pendingFlags) {
		m_pendingFlags = next;
	}
}

void IrqController::acknowledge(uint32_t mask) {
	if (mask != 0u) {
		const uint32_t next = m_pendingFlags & ~mask;
		if (next != m_pendingFlags) {
			m_pendingFlags = next;
		}
	}
	m_memory.writeIoValue(IO_IRQ_ACK, valueNumber(0.0));
}

Value IrqController::onFlagsReadThunk(void* context, uint32_t) {
	const auto* controller = static_cast<IrqController*>(context);
	return valueNumber(static_cast<double>(controller->m_pendingFlags));
}

void IrqController::onAckWriteThunk(void* context, uint32_t, Value value) {
	auto* controller = static_cast<IrqController*>(context);
	controller->acknowledge(toU32(value));
}

} // namespace bmsx
